theano/tensor/signal/downsample.py

```diff
@@ -1146,3 +1159,206 @@ def local_average_pool_grad(node):
+    """
+
+    __props__ = ('ds', 'ignore_border', 'st', 'padding', 'mode')
+
+    @staticmethod
+    def out_shape(imgshape, ds, ignore_border=False):
+        """Return the shape of the output from this op, for input of given shape and flags.
+
+        :param imgshape: the shape of a tensor of images. The last two elements are interpreted
+        as the number of rows, and the number of cols.
+        :type imgshape: tuple, list, or similar.
+
+        :param ds: downsample factor over rows and columns
+        :type ds: list or tuple of two ints
+
+        :param ignore_border: if ds doesn't divide imgshape, do we include an extra row/col of
+        partial downsampling (False) or ignore it (True).
+        :type ignore_border: bool
+
+        :rtype: list
+        :returns: the shape of the output from this op, for input of given shape.  This will
+        have the same length as imgshape, but with last two elements reduced as per the
+        downsampling & ignore_border flags.
+        """
+        if len(imgshape) < 2:
+            raise TypeError('imgshape must have at least two elements (rows, cols)')
+        r, c = imgshape[-2:]
+        rval = list(imgshape[:-2]) + [ r / ds[0], c / ds[1]]
+        if not ignore_border:
+            if r % ds[0]:
+                rval[-2] += 1
+            if c % ds[1]:
+                rval[-1] += 1
+        return rval
+
+    def __init__(self, ds, ignore_border, st=None, padding=(0, 0), mode='max'):
+        """
+        :param ds: downsample factor over rows and columns
+        :type ds: list or tuple of two ints
+
+        :param ignore_border: if ds doesn't divide imgshape, do we include an extra row/col of
+        partial downsampling (False) or ignore it (True).
+        :type ignore_border: bool
+
+        TODO: why is poolsize an op parameter here?
+        """
+        self.ds = tuple(ds)
+        self.ignore_border = ignore_border
+        if st is None:
+            st = ds
+        self.st = tuple(st)
+        self.padding = tuple(padding)
+        self.mode = mode
+        assert self.mode == 'max'
+
+
+    def __eq__(self, other):
+        return type(self) == type(other) and self.ds == other.ds and self.ignore_border == other.ignore_border
+
+    def __hash__(self):
+        return hash(type(self)) ^ hash(self.ds) ^ hash(self.ignore_border)
+
+    def __str__(self):
+        return '%s{%s,%s}' % (self.__class__.__name__, self.ds, self.ignore_border)
+
+    def make_node(self, x, eval_point):
+        if x.type.ndim != 4:
+            raise TypeError('Expected tensor4')
+        if x.type.ndim != 4:
+            return TypeError('Expected tensor4')
+        # TODO: consider restrucing the dtype?
+        x = tensor.as_tensor_variable(x)
+        eval_point = tensor.as_tensor_variable(eval_point)
+        return gof.Apply(self, [x, eval_point], [x.type()])
+
+    def perform(self, node, inp, out):
+        x, ex = inp
+        z, = out
+        if len(x.shape) != 4:
```

Comment:
You need to update this method to do the right implementation when padding != 0 and when strides != ds. If not, raise an error in the __init__ for the not implemented case.

This is python code, the not efficient version. If it help you, you can explicitly pad the inputs if that help you make the support for this case.